File: source/materials/metallic.cpp
#include "materials/metallic.hpp"
#include "math/details.hpp"

#include <algorithm>
#include <cmath>

metallic::metallic(
   const colour& emission_in, const colour& diffuse_in, double roughness_in) noexcept :
   emission(emission_in),
   diffuse(diffuse_in), roughness(std::clamp(roughness_in, 0.0, 1.0))
{}

auto metallic::scatter(const ray& ray_in, const hit& hit_in, [[maybe_unused]] double u,
   [[maybe_unused]] double v) const -> scatter_data
{
   const norm reflected_dir = reflect(hit_in.normal, normalise(ray_in.direction()));
   const ray scattered = ray(hit_in.position, reflected_dir + roughness * random_in_unit_sphere());

   // clang-format off
   return scatter_data {
      .emission = emission, 
      .diffuse = dot(scattered.direction(), hit_in.normal) > 0 ? diffuse : colour{}, 
      .scattered_ray = scattered
   };
   // clang-format on
}

File: go/materials/metallic.go
package materials

import (
	"go_pt/maths"
	"go_pt/tracer"
)

type Metallic struct {
   emissionColour maths.Vec3
   diffuseColour maths.Vec3
   roughness float64
}

func NewMetallicMaterial(emissionColour, diffuseColour *maths.Vec3, roughness float64) Metallic {
   return Metallic{
      emissionColour: *emissionColour,
      diffuseColour: *diffuseColour,
      roughness: roughness} 
}

func (mat *Metallic) Scatter(ray *tracer.Ray, hit *tracer.Hit, u, v float64) MaterialScatterData {
   normalisedRayDir := maths.Normalise(&ray.Direction)
   reflectedDir := maths.Reflect(&hit.Normal, &normalisedRayDir)

   randomDir := maths.RandomVec3InUnitSphere()

   scatteredRay := tracer.Ray{
      Origin: hit.Position, 
      Direction: maths.AddCpy(
         reflectedDir, 
         maths.MultScalar(&randomDir, mat.roughness))}

   if maths.Dot(&scatteredRay.Direction, &hit.Normal) > 0 {
      return MaterialScatterData{
         Emission: mat.emissionColour, 
         Diffuse: mat.diffuseColour,
         ScatteredRay: scatteredRay}
   } else {
      return MaterialScatterData{
         Emission: mat.emissionColour, 
         Diffuse: maths.Vec3{},
         ScatteredRay: scatteredRay}
   }
}
